src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst

```diff
@@ -286,112 +285,106 @@ let send_authenticated_proof #invs #a tr higher_layer_preds comm_keys_ids sender
       ()
     )
 
 
-#push-options "--z3rlimit 50"
 val verify_message_proof:
   {|cinvs:crypto_invariants|} ->
-  #a:Type -> {| ps:parseable_serializeable bytes a |} ->
+  #a:Type -> {|config:comm_layer_core_config a|} ->
   tr:trace ->
   sender:principal -> receiver:principal ->
   msg_bytes:bytes -> sk_receiver_opt:option bytes ->
   vk_sender:bytes ->
   Lemma
   (requires
-    has_communication_layer_crypto_predicates /\
+    has_communication_layer_core_crypto_predicates a /\
     is_publishable tr msg_bytes /\
-    is_public_key_for tr vk_sender (LongTermSigKey comm_layer_sign_tag) sender
+    is_public_key_for tr vk_sender (LongTermSigKey (comm_layer_sign_tag a)) sender
   )
   (ensures (
     match verify_message #a receiver msg_bytes sk_receiver_opt vk_sender with
-    | Some cm -> (
-      let Some msg_signed_t = parse com_message_t msg_bytes in
+    | Some payload -> (
+      let Some msg_signed_t = parse comm_message_t msg_bytes in
       let SigMessage msg_signed = msg_signed_t in
-      let Some sign_input = parse signature_input #parseable_serializeable_bytes_signature_input msg_signed.msg in
-      let payload_bytes = serialize a cm.payload in
-      is_well_formed a (is_publishable tr) cm.payload /\
+      let Some sign_input = parse signature_input msg_signed.msg in
       (
-        (
-          cm.sender == sender /\
-          cm.receiver == receiver /\
-          (match sign_input with
-            | Plain sender' receiver' payload_bytes' -> (
-              event_triggered tr sender (CommAuthSendMsg sender payload_bytes)
-            )
-            | Encrypted sender' receiver' payload_bytes' pk_receiver -> (
-              match parse com_send_byte payload_bytes with
-              | None -> True
-              | Some payload -> (
-                exists plain_payload nonce.
-                  payload.b == pke_enc pk_receiver nonce plain_payload /\
-                  event_triggered tr sender (CommConfAuthSendMsg sender receiver plain_payload)
-                )
+        match sign_input with
+        | Plain sender' receiver' payload_bytes' -> (
+          is_well_formed a (is_publishable tr) (Inl?.v payload) /\
+          (
+            event_triggered tr sender (CommAuthSendMsg sender (Inl?.v payload) <: communication_core_event a) \/ 
+            is_corrupt tr (long_term_key_label sender)
+          )
+        )
+        | Encrypted sender' receiver' payload_bytes' pk_receiver -> (
+          is_publishable tr (Inr?.v payload) /\
+          (
+            (
+              exists plain_payload nonce.
+                (Inr?.v payload) == pke_enc pk_receiver nonce plain_payload /\
+                (match parse a plain_payload with
+                | None -> False
+                | Some plain_payload_parsed -> event_triggered tr sender (CommConfAuthSendMsg sender receiver plain_payload_parsed <: communication_core_event a))
+            ) \/ (
+              is_corrupt tr (long_term_key_label sender)
```

Comment:
Is it expected that we lose the sender/sender' connection here? Previously we had `cm.sender == sender` and similar for the receiver.)

It also seems like `is_corrupt tr (long_term_key_label sender)` could be commuted outside of the match, which saves a few lines, if nothing else.

Reply:
I reinserted the sender/sender' and receiver/receiver' connection.

I cannot pull out the `is_corrupt tr (long_term_key_label sender)` because the `is_publishable` also has to hold if the sender is corrupt, and I am not sure how to cleanly pull that out with the `either` type used for the payload.